### PR TITLE
Fix telemetry method and object strings

### LINKED ISSUE #6866: Invariant Violation: "Method "discoverynavigation" must be one of the registered values
### Describe the problem and steps to reproduce it:

An event that logs tracking events when clicking the "Find more add-ons button" isn't registered and causes an invariant violation.

```
Unhandled promise rejection Invariant Violation: "Method "discoverynavigation" must be one of the registered values: addonthemeclicks,addonactivation,themeactivation,addoninstallscancell,themeinstallscancell,addoninstalls,addoninstallsstarted,themeinstallsstarted,themeinstalls,addonuninstalls,themeuninstalls"
```

Refs: #6797

### LINKED ISSUE #6867: Unhandled promise rejection Invariant Violation: "Method "addoninstallsdownloa" must be one of the registered values:
### Describe the problem and steps to reproduce it:

An event that logs an event for `AMO Addon Installs Download Failed` isn't registered and causes an invariant violation.

Another problem here is that due to string limitations the meaning of this message is hidden.

```
Unhandled promise rejection Invariant Violation: "Method "addoninstallsdownloa" must be one of the registered values: addonthemeclicks,addonactivation,themeactivation,addoninstallscancell,themeinstallscancell,addoninstalls,addoninstallsstarted,themeinstallsstarted,themeinstalls,addonuninstalls,themeuninstalls"
```

I think to fix this the registration of the events needs to be improved as well as covering the possibility of this specific event being logged.

Refs: #6797

### PR DESCRIPTION
**Pull request by muffinresearch**
Fixes #6867 
Fixes #6866 

This patch cleans-up the strings used for HCT - because an issue with filtering strings is that it was causing the strings to be truncated which in some cases hid their meaning. 

To fix this I've removed the previous filtering in favour of a fixed mapping for the methods so readability is ensured. The tests check that the mapping text meets the [HCT requirements](https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/collection/events.html#limits).

At the same time as implementing better strings, I've also documented the data collection and added a test to ensure we always document any additional methods.

Results now look like this:

telemetry_data

### CODE REVIEW

**Comment by muffinresearch** [comment]
One remaining question I have, there's quite some duplication in the data in terms of events to add-ons or themes, should we look to remove that? 

I'm thinking now would probably be a better time to do that rather than later.

One possible fix would be to move `ext_installed` and `theme_installed` to be one string: `addon_install`. This won't impact existing GA collection now and the way this is set-up changing those constants later shouldn't impact the telemetry collection.

File: tests/unit/core/test_tracking.js [hunk 1]
```diff
@@ -432,20 +434,22 @@ describe(__filename, () => {
...
-  describe('Hybrid Content Telemetry filterIdentifier()', () => {
-    it('should return content up to the default max length', () => {
-      expect(filterIdentifier('a'.repeat(26))).toHaveLength(20);
-    });
+  describe('HCT identifiers', () => {
+    const telemetryRegex = /^[a-z0-9]{1}[a-z0-9_]+[a-z0-9]{1}$/i;
 
-    it('should return content up to the supplied max length', () => {
-      expect(filterIdentifier('a'.repeat(26), { maxLen: 10 })).toHaveLength(10);
+    Object.values(HCT_METHOD_MAPPING).forEach((idString) => {
```
Comment (willdurand) [inline tests/unit/core/test_tracking.js]: Jest introduced a `.each` method on the `it` and `describe` functions. We tend to use it:

```diff
diff --git a/tests/unit/core/test_tracking.js b/tests/unit/core/test_tracking.js
index 3c785a1a9..68e98d143 100644
--- a/tests/unit/core/test_tracking.js
+++ b/tests/unit/core/test_tracking.js
@@ -437,12 +437,13 @@ describe(__filename, () => {
   describe('HCT identifiers', () => {
     const telemetryRegex = /^[a-z0-9]{1}[a-z0-9_]+[a-z0-9]{1}$/i;
 
-    Object.values(HCT_METHOD_MAPPING).forEach((idString) => {
-      it(`should ensure hct method "${idString}" meets HCT identifier requirements`, () => {
+    it.each(Object.values(HCT_METHOD_MAPPING))(
+      'should ensure hct method "%s" meets HCT identifier requirements',
+      (idString) => {
         expect(idString).toMatch(telemetryRegex);
         expect(idString.length).toBeLessThanOrEqual(20);
-      });
-    });
+      },
+    );
 
     for (const action of telemetryObjects) {
       it(`should ensure hct object (${action}) meets HCT identifier requirements`, () => {
```

The function gets the `idString` as first argument and you use `%s` in the test description.

File: tests/unit/core/test_tracking.js [hunk 2]
```diff
@@ -432,20 +434,22 @@ describe(__filename, () => {
...
+        expect(idString.length).toBeLessThanOrEqual(20);
+      });
     });
 
-    it('should return content filtered for telemetry', () => {
-      expect(filterIdentifier('a/bcde/-#/._fgh', { maxLen: 10 })).toEqual(
-        'abcde_fgh',
-      );
-    });
+    for (const action of telemetryObjects) {
```
Comment (willdurand) [inline tests/unit/core/test_tracking.js]: Same here:

```diff
diff --git a/tests/unit/core/test_tracking.js b/tests/unit/core/test_tracking.js
index 3c785a1a9..549ccc1f6 100644
--- a/tests/unit/core/test_tracking.js
+++ b/tests/unit/core/test_tracking.js
@@ -444,12 +444,13 @@ describe(__filename, () => {
       });
     });
 
-    for (const action of telemetryObjects) {
-      it(`should ensure hct object (${action}) meets HCT identifier requirements`, () => {
+    it.each(telemetryObjects)(
+      'should ensure hct object (%s) meets HCT identifier requirements',
+      (action) => {
         expect(action).toMatch(telemetryRegex);
         expect(action.length).toBeLessThanOrEqual(20);
-      });
-    }
+      },
+    );
   });
 
   describe('Tracking constants should not be changed or it risks breaking tracking stats', () => {
```

File: tests/unit/test_tracking_docs.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { readFile } from 'fs';
+
+import { HCT_METHOD_MAPPING } from 'core/constants';
+
+describe(__filename, () => {
+  describe('Tracking docs', () => {
```
Comment (willdurand) [inline tests/unit/test_tracking_docs.js]: ❤️

**Comment by willdurand** [comment]
> At the same time as implementing better strings, I've also documented the data collection and added a test to ensure we always document any additional methods.

Can we document how to setup HCT for local development? Here are the steps I followed:

- turn `hctEnabled` ON in `config/development-disco.js`. We should probably commit that (here?)
- run `yarn disco:https` because we need HTTPS
- open the Browser Console (and not the classic devtools) and type:

```js
const hostURI = Services.io.newURI('https://example.com:3000');
Services.perms.add(hostURI, "hc_telemetry", Services.perms.ALLOW_ACTION);
```

- Browse `about:telemetry#events-tab` and select the "dynamic" filter (top-right dropdown)

**Comment by willdurand** [comment]
> @willdurand One remaining question I have, there's quite some duplication in the data in terms of events to add-ons or themes, I wondering it would make sense remove that? Now would probably be a better time to do that rather than later.
> 
> One possible fix would be to move `ext_installed` and `theme_installed` to be one string: `addon_install`. This won't impact existing GA collection now and the way this is set-up changing those constants later shouldn't impact the telemetry collection.
> 
> If that sounds good I'll file a follow-up patch.

I am not quite sure to know why we have this distinction in GA (I find it weird that it is "addon" vs "theme", with "addon" gathering different add-on types). That being said, it looks like for HCT, we can merge most "methods" since "objects" describe the add-on types, so we can still retrieve the breakdown like in GA.

File: tests/unit/core/test_tracking.js (hunk 1 above)
Comment (muffinresearch) [reply tests/unit/core/test_tracking.js]: Yep that's how `it.each` works, those are positional placeholders.

Comment (muffinresearch) [reply tests/unit/core/test_tracking.js]: See https://jestjs.io/docs/en/api#1-testeachtable-name-fn-timeout

**Comment by muffinresearch** [comment]
> I am not quite sure to know why we have this distinction in GA (I find it weird that it is "addon" vs "theme", with "addon" gathering different add-on types). That being said, it looks like for HCT, we can merge most "methods" since "objects" describe the add-on types, so we can still retrieve the breakdown like in GA.

You and I both :). I'd like to see that factored out at some point, but we need to ensure continuity of tracking. I'll file an issue for the GA case.

**Comment by muffinresearch** [comment]
@willdurand ok I've added a second commit that adds testing documentation and de-dupes the HCT method strings. Could you take another look?

File: src/core/constants.js [hunk 4]
```diff
@@ -236,6 +236,38 @@ export const SURVEY_ACTION_DISMISSED = 'Dismissed survey notice';
...
+  [UNINSTALL_EXTENSION_CATEGORY]: HCT_ADDON_UNINSTALLED,
+  [UNINSTALL_THEME_CATEGORY]: HCT_ADDON_UNINSTALLED,
+  [ENABLE_EXTENSION_CATEGORY]: HCT_ADDON_ENABLED,
+  [ENABLE_THEME_CATEGORY]: HCT_ADDON_ENABLED,
+  [INSTALL_CANCELLED_EXTENSION_CATEGORY]: HCT_ADDON_INSTALL_CANCELLED,
+  [INSTALL_CANCELLED_THEME_CATEGORY]: HCT_ADDON_INSTALL_CANCELLED,
+  [INSTALL_STARTED_EXTENSION_CATEGORY]: HCT_ADDON_INSTALL_STARTED,
+  [INSTALL_STARTED_THEME_CATEGORY]: HCT_ADDON_INSTALL_STARTED,
+  [INSTALL_DOWNLOAD_FAILED_EXTENSION_CATEGORY]: HCT_ADDON_DOWNLOAD_FAILED,
+  [INSTALL_DOWNLOAD_FAILED_THEME_CATEGORY]: HCT_ADDON_DOWNLOAD_FAILED,
```
Comment (willdurand) [inline src/core/constants.js]: Would it make sense to add tests for this mapping?

Comment (muffinresearch) [reply src/core/constants.js]: In what sense?

File: tests/unit/core/test_tracking.js (hunk 1 above)
Comment (muffinresearch) [reply tests/unit/core/test_tracking.js]: Hmm I clearly got a bit confused here - I somehow thought you were commenting on my use of `it.each` elsewhere. Happy to move to using it here too. I used it in the doc tests already.

File: src/core/constants.js (hunk 4 above)
Comment (willdurand) [reply src/core/constants.js]: Making sure that both `INSTALL_DOWNLOAD_FAILED_THEME_CATEGORY` and `INSTALL_DOWNLOAD_FAILED_EXTENSION_CATEGORY` map to `HCT_ADDON_DOWNLOAD_FAILED`, etc.?

File: tests/unit/core/test_tracking.js (hunk 1 above)
Comment (muffinresearch) [reply tests/unit/core/test_tracking.js]: My bad! That's all fixed now.

**Pull request merged**